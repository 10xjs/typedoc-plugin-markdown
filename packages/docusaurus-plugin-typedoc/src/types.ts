export interface PluginOptions {
  inputFiles: string[];
  docsRoot: string;
  out: string;
  sidebar: SidebarOptions | null;
  plugin: string[];
  disableOutputCheck?: boolean;
}

export interface SidebarOptions {
  fullNames: boolean;
  sidebarFile: string;
  globalsLabel: string;
  readmeLabel: string;
}

export interface Sidebar {
  [sidebarId: string]: SidebarItem[];
}

export interface SidebarCategory {
  type: string;
  label: string;
  items: SidebarItem[];
}

export type SidebarItem = SidebarCategory | string;

export interface LoadedContent {
  navigation: any;
}
