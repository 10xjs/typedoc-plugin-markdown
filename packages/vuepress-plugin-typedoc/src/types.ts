import { Application, ProjectReflection } from 'typedoc';

export interface PluginOptions {
  id?: string;
  logger?: string;
  inputFiles?: string[];
  out?: string;
  hideBreadcrumbs: boolean;
  hideProjectName: boolean;
  sidebar?: SidebarOptions | null;
}

export interface SidebarOptions {
  fullNames: boolean;
  parentCategory: string;
}

export interface LoadedContent {
  app: Application;
  project: ProjectReflection;
}
