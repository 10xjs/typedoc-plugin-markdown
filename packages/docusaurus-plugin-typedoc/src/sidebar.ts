import * as path from 'path';

import * as fs from 'fs-extra';
import { NavigationItem } from 'typedoc';

import { Sidebar, SidebarItem, SidebarOptions } from './types';

export function writeSidebar(
  siteDir: string,
  outFolder: string,
  sidebar: SidebarOptions,
  navigation: NavigationItem,
) {
  // map the navigation object to a Docuaurus sidebar format
  const sidebarItems = navigation.children
    ? navigation.children.map((navigationItem) => {
        if (navigationItem.url && navigationItem.children?.length === 0) {
          return getUrlKey(outFolder, navigationItem.url);
        } else {
          const sidebarCategoryItems = navigationItem.children
            ? navigationItem.children.map((navItem) => {
                const url = getUrlKey(outFolder, navItem.url);
                if (navItem.children && navItem.children.length > 0) {
                  const sidebarCategoryChildren = navItem.children.map(
                    (childGroup) =>
                      sidebarCategory(
                        childGroup.title,
                        childGroup.children
                          ? childGroup.children.map((childItem) =>
                              getUrlKey(outFolder, childItem.url),
                            )
                          : [],
                      ),
                  );
                  return sidebarCategory(navItem.title, [
                    url,
                    ...sidebarCategoryChildren,
                  ]);
                }
                return url;
              })
            : [];
          return sidebarCategory(navigationItem.title, sidebarCategoryItems);
        }
      })
    : [];

  // full path to sidebar
  const sidebarPath = path.resolve(siteDir, sidebar.sidebarFile);

  // if sidebar already exists merge non typedoc sidebars
  const initialSidebarContent = fs.existsSync(sidebarPath)
    ? require(sidebarPath)
    : JSON.parse('{}');

  const sidebarContent: Sidebar = Object.assign({}, initialSidebarContent, {
    typedoc:
      sidebar.parentCategory && sidebar.parentCategory !== 'none'
        ? [
            {
              type: 'category',
              label: sidebar.parentCategory,
              items: sidebarItems,
            },
          ]
        : sidebarItems,
  });

  // write result to disk
  fs.outputFileSync(
    sidebarPath,
    'module.exports = ' + JSON.stringify(sidebarContent, null, 2) + ';',
  );

  console.log(`[docusaurus-plugin-typedoc] sidebar written to ${sidebarPath}`);
}

/**
 * returns a sidebar category node
 */
function sidebarCategory(title: string, items: SidebarItem[]) {
  return {
    type: 'category',
    label: title,
    items,
  };
}

/**
 * returns the url key for relevant doc
 */
function getUrlKey(outFolder: string, url: string) {
  const urlKey = url.replace('.md', '');
  return outFolder ? outFolder + '/' + urlKey : urlKey;
}
