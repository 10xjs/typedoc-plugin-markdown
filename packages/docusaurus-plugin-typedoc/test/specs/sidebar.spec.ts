import * as fs from 'fs';
import * as path from 'path';

import * as tmp from 'tmp';

import { TestApp } from '../../../typedoc-plugin-markdown/test/test-app';
import { writeSidebar } from '../../dist/sidebar';
import { SidebarOptions } from '../../dist/types';

tmp.setGracefulCleanup();

describe(`Sidebar:`, () => {
  let testApp: TestApp;

  beforeAll(() => {
    testApp = new TestApp(['theme.ts']);
    testApp.bootstrap({
      theme: path.resolve(__dirname, '..', '..', 'dist', 'theme'),
    });
  });

  test(`should merge existing sidebar with a parent category`, () => {
    const tmpobj = tmp.dirSync();
    const navigation = testApp.theme.getNavigation(testApp.project);
    const sidebarFile = 'sidebars.js';
    const sidebarOptions: SidebarOptions = {
      fullNames: false,
      parentCategory: 'API',
      sidebarFile,
    };
    const sidebarsContent = `module.exports = {
      "someSidebar": {
        "Docusaurus": [
          "doc1",
        ],
        "Features": [
          "mdx"
        ]
      }
    };`;
    fs.writeFileSync(path.resolve(tmpobj.name, sidebarFile), sidebarsContent);
    writeSidebar(tmpobj.name, 'api', sidebarOptions, navigation);
    const sidebar = fs.readFileSync(path.resolve(tmpobj.name, sidebarFile));
    expect(sidebar.toString()).toMatchSnapshot();
  });

  test(`should write to a new sidebar with custom id and root docs`, () => {
    const tmpobj = tmp.dirSync();
    const navigation = testApp.theme.getNavigation(testApp.project);
    const sidebarFile = 'sidebars/newSidebar.js';
    const sidebarOptions: SidebarOptions = {
      fullNames: false,
      parentCategory: 'none',
      sidebarFile,
    };
    writeSidebar(tmpobj.name, '', sidebarOptions, navigation);
    const sidebar = fs.readFileSync(path.resolve(tmpobj.name, sidebarFile));
    expect(sidebar.toString()).toMatchSnapshot();
  });
});
