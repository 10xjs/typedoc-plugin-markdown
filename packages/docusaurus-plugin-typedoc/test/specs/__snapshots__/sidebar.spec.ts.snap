// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar: should merge existing sidebar with a parent category 1`] = `
"module.exports = {
  \\"someSidebar\\": {
    \\"Docusaurus\\": [
      \\"doc1\\"
    ],
    \\"Features\\": [
      \\"mdx\\"
    ]
  },
  \\"typedoc\\": [
    {
      \\"type\\": \\"category\\",
      \\"label\\": \\"API\\",
      \\"items\\": [
        \\"api/index\\",
        \\"api/globals\\",
        {
          \\"type\\": \\"category\\",
          \\"label\\": \\"Modules\\",
          \\"items\\": [
            \\"api/modules/_theme_\\"
          ]
        },
        {
          \\"type\\": \\"category\\",
          \\"label\\": \\"Classes\\",
          \\"items\\": [
            \\"api/classes/_theme_.classa\\",
            \\"api/classes/_theme_.classb\\"
          ]
        },
        {
          \\"type\\": \\"category\\",
          \\"label\\": \\"Interfaces\\",
          \\"items\\": [
            \\"api/interfaces/_theme_.interfacea\\",
            \\"api/interfaces/_theme_.interfaceb\\"
          ]
        }
      ]
    }
  ]
};"
`;

exports[`Sidebar: should write to a new sidebar with custom id and root docs 1`] = `
"module.exports = {
  \\"typedoc\\": [
    \\"index\\",
    \\"globals\\",
    {
      \\"type\\": \\"category\\",
      \\"label\\": \\"Modules\\",
      \\"items\\": [
        \\"modules/_theme_\\"
      ]
    },
    {
      \\"type\\": \\"category\\",
      \\"label\\": \\"Classes\\",
      \\"items\\": [
        \\"classes/_theme_.classa\\",
        \\"classes/_theme_.classb\\"
      ]
    },
    {
      \\"type\\": \\"category\\",
      \\"label\\": \\"Interfaces\\",
      \\"items\\": [
        \\"interfaces/_theme_.interfacea\\",
        \\"interfaces/_theme_.interfaceb\\"
      ]
    }
  ]
};"
`;
